Sources/SourceKitLSP/Swift/CodeActions/ApplyDeMorganLaw.swift

```diff
@@ -0,0 +1,730 @@
+          self.init(kind: .and, exprType: .bitwise)
+        case .binaryOperator("||"):
+          self.init(kind: .or, exprType: .boolean)
+        case .binaryOperator("&&"):
+          self.init(kind: .and, exprType: .boolean)
+        default:
+          return nil
+        }
+      }
+
+      var complement: Self {
+        Self(kind: self.kind.complement, exprType: self.exprType)
+      }
+
+      var tokenKind: TokenKind {
+        switch (self.kind, self.exprType) {
+        case (.and, .boolean):
+          .binaryOperator("&&")
+        case (.or, .boolean):
+          .binaryOperator("||")
+        case (.and, .bitwise):
+          .binaryOperator("&")
+        case (.or, .bitwise):
+          .binaryOperator("|")
+        }
+      }
+    }
+
+    /// The binary operator of this negated expression.
+    var `operator`: Operator
+    /// The negated expression.
+    var expr: InfixOperatorExprSyntax
+
+    /// Constructs the negated expression of the given `infixExpr` if it is propositions.
+    ///
+    /// If `exprType` is `nil`, this indicates `infixExpr` is at the root-level and `exprType` will be inferred from
+    /// the type of the binary operator.
+    /// - Note: The form `a AND/OR b /* negated to */ NOT a OR/AND NOT b` is considered unproductive at the root-level
+    /// thus not supported, e.g. `a || b /* negated to */ !a && !b`.
+    ///
+    /// Otherwise, `infixExpr` will be checked for its binary operator's agreement with the given `exprType`.
+    ///
+    /// To preserve precedence order, a subexpression of the negated expression will be enclosed with an extra pair of
+    /// parentheses if,
+    ///   1. The negated expression has flipped from OR to AND, and
+    ///   2. The subexpression has flipped from AND to OR, and
+    ///   3. The subexpression is not `TupleExprSyntax`.
+    ///
+    /// The necessity of such parenthesising can be seen as follows,
+    /// ```swift
+    /// a && b || c /* negated to */ (!a || !b) && !c
+    /// ```
+    init?(_ infixExpr: InfixOperatorExprSyntax, exprType: ExprType? = nil) {
+      let biOperator = infixExpr.biOperator
+      guard let `operator` = Operator(biOperator.operator.tokenKind)?.complement else {
+        return nil
+      }
+      self.operator = `operator`
+
+      func exprParenthesizedIfNeeded(_ negated: Expr.Negated) -> ExprSyntax {
+        if `operator`.kind == .and && negated.change == .andToOr && !negated.expr.is(TupleExprSyntax.self) {
+          ExprSyntax(TupleExprSyntax(onlyExpr: negated.expr))
+        } else {
+          negated.expr
+        }
+      }
+
+      let leftNegated: Expr.Negated
+      let rightNegated: Expr.Negated
+      if let exprType {
+        guard self.operator.exprType == exprType else {
+          return nil
+        }
+        leftNegated = Expr(expr: infixExpr.leftOperand).negated(exprType: self.operator.exprType)
+        rightNegated = Expr(expr: infixExpr.rightOperand).negated(exprType: self.operator.exprType)
```

Comment:
I think this is the same in both `if` branches, so we should be able to extract it to the definition of `leftNegated`.